### DISCUSSION TITLE
Script doesn't run on import but does run when I trigger manually

### DISCUSSION THREAD

**Discussion by njproductionsus** [AntlerVC/firetable#1495] [Support / Q&A]
I am testing a simple scenario for something and am stumped by these results:

1. My script in **testoutput** listens for a number of 1, 2, or 3 in **testdelay#**.
2. The script is designed so that if a 1, it outputs immediately; if a 2, it outputs in 60 seconds; if a 3, it outputs in 120 seconds.
3. However, when I import rows in, it does NOT output for numbers 2 or 3 (the ones with a delay), only for 1.
4. If I then clear each output and rerun the ENTIRE column, the 2 and 3 DOES work in 60/120 seconds accordingly.

SO, that brings me to my question - why does 2/3 not work on import but it does when manually run?

Here is the code for **testoutput** (note logs are also matching the results and no errors being noted):

const derivative: Derivative = async ({ row, ref, db, storage, auth, logging }) => {
  try {    
      // Add a delay based on the value in the 'testdelay' column
      const delayValue = row.testdelay;
      if (delayValue === 1) {
        await ref.update({ testoutput: "Ran with no delay." });
        logging.log("Ran with no delay.");
      } else if (delayValue === 2) {
        await new Promise((resolve) => setTimeout(resolve, 60000)); // 60 seconds
        await ref.update({ testoutput: "Ran with a 60-second delay." });
        logging.log("Ran with a 60-second delay.");
      } else if (delayValue === 3) {
        await new Promise((resolve) => setTimeout(resolve, 120000)); // 120 seconds
        await ref.update({ testoutput: "Ran with a 120-second delay." });
        logging.log("Ran with a 120-second delay.");
      } else {
        await ref.update({ testoutput: "Invalid delay value. No delay applied." });
        logging.log("Invalid delay value. No delay applied.");
      }    
  } catch (error) {
    await ref.update({ testoutput: "An error occurred: " + error.message });
    logging.error("An error occurred: " + error.message);
  }
};

**Comment by njproductionsus** [comment]
Is it possible that on IMPORT - there is a timeout issue that is stopping the script to wait 60 seconds / 120 seconds? 

But when I MANUALLY run the same script - there is no such timeout?

**Comment by Chadha93** [comment]
@njproductionsus When you import rows, the script runs almost instantaneously, which means that any delays specified in the code may not have enough time to execute before the script completes.

One possible solution could be to modify your code to handle the delays outside of the Cloud Function. Instead of using `setTimeout` within the function, you can add a separate step to schedule the delayed updates using a Cloud Scheduler or a Cloud Pub/Sub topic.

Here's an example of how you can modify your code to achieve this:

```javascript
const derivative: Derivative = async ({ row, ref, db, storage, auth, logging }) => {
  try {    
    const delayValue = row.testdelay;
    if (delayValue === 1) {
      await ref.update({ testoutput: "Ran with no delay." });
      logging.log("Ran with no delay.");
    } else if (delayValue === 2) {
      // Schedule a delayed update using Cloud Scheduler or Cloud Pub/Sub
      await scheduleDelayedUpdate(ref, "Ran with a 60-second delay.", 60);
    } else if (delayValue === 3) {
      // Schedule a delayed update using Cloud Scheduler or Cloud Pub/Sub
      await scheduleDelayedUpdate(ref, "Ran with a 120-second delay.", 120);
    } else {
      await ref.update({ testoutput: "Invalid delay value. No delay applied." });
      logging.log("Invalid delay value. No delay applied.");
    }    
  } catch (error) {
    await ref.update({ testoutput: "An error occurred: " + error.message });
    logging.error("An error occurred: " + error.message);
  }
};

const scheduleDelayedUpdate = async (ref, message, delaySeconds) => {
  // Use Cloud Scheduler or Cloud Pub/Sub to schedule the delayed update
  // Example code for using Cloud Scheduler
  // Replace 'YOUR_PROJECT_ID' with your actual project ID
  const projectId = 'YOUR_PROJECT_ID';
  const scheduler = require('@google-cloud/scheduler');
  const client = new scheduler.CloudSchedulerClient();
  const parent = client.locationPath(projectId, 'us-central1');

  const job = {
    name: `delayed-update-${Date.now()}`,
    schedule: `*/${delaySeconds} * * * *`, // Run every delaySeconds minutes
    timeZone: 'America/Los_Angeles',
    httpTarget: {
      uri: `https://${projectId}.cloudfunctions.net/updateFunction?ref=${ref.path}&message=${message}`,
      httpMethod: 'POST',
    },
  };

  await client.createJob({
    parent: parent,
    job: job,
  });
};
```

**Reply by njproductionsus** [reply]
Thank you for this detailed example! I will test this out but it does help to explain the issue in the first place (as far as Rowy on import is processing all too quickly to have the delay with my method).
